### PR TITLE
[flang]Add vscale argument parsing

### PR DESCRIPTION
**Pull request by Leporacanthicus**
Support for vector scale range arguments, for AArch64 scalable vector extension (SVE) support.

Adds -msve-vector-bits to the flang frontend, and for flang fc1 the options are -mvscale-min and -mvscale-max (optional). These match the clang and clang cc1 options for the same purposes.

A further patch will actually USE these arguments.

### CODE REVIEW

File: flang/lib/Frontend/FrontendActions.cpp [hunk 1]
```diff
@@ -42,6 +42,7 @@
 #include "mlir/Target/LLVMIR/ModuleTranslation.h"
 #include "clang/Basic/Diagnostic.h"
 #include "clang/Basic/DiagnosticFrontend.h"
+#include "clang/Basic/TargetInfo.h"
```
Comment (tblah) [inline flang/lib/Frontend/FrontendActions.cpp:45]: Is this used anywhere?

**Review by tblah — CHANGES_REQUESTED** [review]
Please could you add testing, like the tests you wrote for the loop versioning flag.

File: clang/lib/Driver/ToolChains/Flang.cpp [hunk 2]
```diff
@@ -170,6 +170,39 @@ void Flang::addPicOptions(const ArgList &Args, ArgStringList &CmdArgs) const {
...
+      }
+
+      bool Invalid = Val.getAsInteger(10, Bits); (void)Invalid;
+      assert(!Invalid && "Failed to parse value");
+      CmdArgs.push_back(
+          Args.MakeArgString("-mvscale-min=" + llvm::Twine(Bits / 128)));
+    // Silently drop requests for vector-length agnostic code as it's implied.
+    } else if (!Val.equals("scalable"))
+      // Handle the unsupported values passed to msve-vector-bits.
+      D.Diag(diag::err_drv_unsupported_option_argument)
```
[the remark this answers is not in the document]
Comment (Leporacanthicus) [reply clang/lib/Driver/ToolChains/Flang.cpp:201]: It's not "unsupported", it just doesn't do anything different from "not specifying anything". It's what clang does.

**Comment by Leporacanthicus** [comment]
> Please could you add testing, like the tests you wrote for the loop versioning flag.

Tests added.

**Review by DavidTruby — APPROVED** [review]
LGTM :)

File: clang/lib/Driver/ToolChains/Flang.cpp [hunk 3]
```diff
@@ -170,6 +169,38 @@ void Flang::addPicOptions(const ArgList &Args, ArgStringList &CmdArgs) const {
...
+  if (Arg *A = Args.getLastArg(options::OPT_msve_vector_bits_EQ)) {
+    StringRef Val = A->getValue();
+    const Driver &D = getToolChain().getDriver();
+    if (Val.equals("128") || Val.equals("256") || Val.equals("512") ||
+        Val.equals("1024") || Val.equals("2048") || Val.equals("128+") ||
+        Val.equals("256+") || Val.equals("512+") || Val.equals("1024+") ||
+        Val.equals("2048+")) {
+      unsigned Bits = 0;
+      if (Val.endswith("+"))
+        Val = Val.substr(0, Val.size() - 1);
```
[the remark this answers is not in the document]
Comment (Leporacanthicus) [reply clang/lib/Driver/ToolChains/Flang.cpp:184]: That's copied directly from clang, and I think it's better to keep them identical [although I did modify the [[maybe_unused]] from (void) in the check if the conversion worked.

[There's about 200 lines AROUND this bit, that does a bunch of other target-specific stuff for aarch64, so I can't just call the clang version of this code]

File: flang/test/Driver/aarch64-sve-vector-bits.f90 [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+! RUN: %flang -c %s -### --target=aarch64-none-linux-gnu -march=armv8-a+sve \
+! RUN:  -msve-vector-bits=512+ 2>&1 | FileCheck --check-prefix=CHECK-512P %s
+! RUN: %flang -c %s -### --target=aarch64-none-linux-gnu -march=armv8-a+sve \
+! RUN:  -msve-vector-bits=1024+ 2>&1 | FileCheck --check-prefix=CHECK-1024P %s
+! RUN: %flang -c %s -### --target=aarch64-none-linux-gnu -march=armv8-a+sve \
+! RUN:  -msve-vector-bits=2048+ 2>&1 | FileCheck --check-prefix=CHECK-2048P %s
+! RUN: %flang -c %s -### --target=aarch64-none-linux-gnu -march=armv8-a+sve \
+! RUN:  -msve-vector-bits=scalable 2>&1 | FileCheck --check-prefix=CHECK-SCALABLE %s
+
+! CHECK-128: "-mvscale-max=1" "-mvscale-min=1"
```
Comment (banach-space) [inline flang/test/Driver/aarch64-sve-vector-bits.f90:28]: Is this from `flang-new -fc1` invocation? It could be written as:
```
! CHECK-128: "-fc1"
! CHECK-128-SAME: "-mvscale-max-1" "-mvscale-min=1"
```

**Review by banach-space — APPROVED** [review]
LGTM, thanks!

File: flang/test/Driver/aarch64-sve-vector-bits.f90 (hunk 4 above)
Comment (banach-space) [reply flang/test/Driver/aarch64-sve-vector-bits.f90:28]: Let me rephrase - what invocation exactly is this check-line checking?

Comment (Leporacanthicus) [reply flang/test/Driver/aarch64-sve-vector-bits.f90:28]: It is checking the fc1 command. Just like the clang test is checking the cc1 command generated by the driver.

So you're saying we should check that it's not catching some other - preprocessor, linker or something else that may have got hold of the "right" command-line argument?

**Review by tblah — APPROVED** [review]

File: flang/test/Driver/aarch64-sve-vector-bits.f90 (hunk 4 above)
Comment (banach-space) [reply flang/test/Driver/aarch64-sve-vector-bits.f90]: > So you're saying we should check that it's not catching some other - preprocessor, linker or something else that may have got hold of the "right" command-line argument?

Indeed. Also, as a documentation for our future selves. It won't hurt, but could make our lives easier when re-visiting this in e.g. 6 months time :) 

This is a nit, feel free to land as is 👍🏻 Thank you :)
